src/test/java/com/google/firebase/internal/GaeExecutorServiceTest.java

```diff
@@ -0,0 +1,129 @@
+import java.util.concurrent.Future;
+import java.util.concurrent.Semaphore;
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import org.junit.Test;
+
+public class GaeExecutorServiceTest {
+
+  @Test
+  public void testShutdownBeforeUse() throws InterruptedException {
+    CountingThreadFactory threadFactory = new CountingThreadFactory();
+    GaeExecutorService executorService = new GaeExecutorService("test", threadFactory);
+    assertFalse(executorService.isShutdown());
+    assertFalse(executorService.isTerminated());
+
+    assertEquals(ImmutableList.of(), executorService.shutdownNow());
+    assertTrue(executorService.isShutdown());
+    assertTrue(executorService.isTerminated());
+    assertTrue(executorService.awaitTermination(1, TimeUnit.SECONDS));
+    assertEquals(0, threadFactory.counter.get());
+
+    executorService = new GaeExecutorService("test", threadFactory);
+    assertFalse(executorService.isShutdown());
+    assertFalse(executorService.isTerminated());
+
+    executorService.shutdownNow();
+    assertTrue(executorService.isShutdown());
+    assertTrue(executorService.isTerminated());
+    assertTrue(executorService.awaitTermination(
+        TestUtils.TEST_TIMEOUT_MILLIS, TimeUnit.MILLISECONDS));
+    assertEquals(0, threadFactory.counter.get());
+  }
+
+  @Test
+  public void testSubmit() throws InterruptedException, ExecutionException {
+    CountingThreadFactory threadFactory = new CountingThreadFactory();
+    GaeExecutorService executorService = new GaeExecutorService("test", threadFactory);
+
+    final Semaphore semaphore = new Semaphore(0);
+    Future future = executorService.submit(new Runnable() {
+      @Override
+      public void run() {
+        semaphore.release();
+      }
+    });
+    assertNotNull(future);
+    waitFor(semaphore);
+
+    future = executorService.submit(new Runnable() {
+      @Override
+      public void run() {
+        semaphore.release();
+      }
+    }, "result");
+    assertNotNull(future);
+    waitFor(semaphore);
+    assertEquals("result", future.get());
+
+    future = executorService.submit(new Callable() {
+      @Override
+      public Object call() throws Exception {
+        semaphore.release();
+        return "result2";
+      }
+    });
+    assertNotNull(future);
+    waitFor(semaphore);
+    assertEquals("result2", future.get());
+
+    executorService.execute(new Runnable() {
+      @Override
+      public void run() {
+        semaphore.release();
+      }
+    });
+    waitFor(semaphore);
+
+    assertEquals(4, threadFactory.counter.get());
+
+    executorService.shutdown();
+    assertTrue(executorService.awaitTermination(
+        TestUtils.TEST_TIMEOUT_MILLIS, TimeUnit.MILLISECONDS));
+    assertTrue(executorService.isShutdown());
+    assertTrue(executorService.isTerminated());
+  }
+
+  private static class CountingThreadFactory implements ThreadFactory {
```

Comment:
I was hoping this was part of the JDK :/ This seems incredible useful!

Reply:
It seems to be supported by [`BasicThreadFactory`](https://commons.apache.org/proper/commons-lang/javadocs/api-3.0/org/apache/commons/lang3/concurrent/BasicThreadFactory.html#getThreadCount()) of commons-lang. But we are currently not depending on it.